examples/apps/ai-collab/src/components/TaskGroup.tsx

```diff
@@ -59,8 +68,91 @@ export function TaskGroup(props: {
 		newBranchTargetNode: SharedTreeTaskGroup;
 	}>();
 
+	const [undoStack, setUndoStack] = useState<Revertible[]>([]);
+	const [redoStack, setRedoStack] = useState<Revertible[]>([]);
+
 	useSharedTreeRerender({ sharedTreeNode: props.sharedTreeTaskGroup, logId: "TaskGroup" });
 
+	/**
+	 * Create undo and redo stacks of {@link Revertible}.
+	 */
+	useEffect(() => {
+		function onRevertibleDisposed(disposed: Revertible): void {
+			const redoIndex = redoStack.indexOf(disposed);
+			if (redoIndex === -1) {
+				const undoIndex = undoStack.indexOf(disposed);
+				if (undoIndex !== -1) {
+					setUndoStack((currUndoStack) => {
+						const newUndoStack = [...currUndoStack];
+						newUndoStack.splice(undoIndex, 1);
+						return newUndoStack;
+					});
+				}
+			} else {
+				setRedoStack((currRedostack) => {
+					const newRedoStack = [...currRedostack];
+					newRedoStack.splice(redoIndex, 1);
+					return newRedoStack;
+				});
+			}
+		}
+
+		/**
+		 * Instead of application developer manually managing the life cycle of the {@link Revertible} instances,
+		 * example app stores upto `MAX_STACK_SIZE` number of {@link Revertible} instances in the undo and redo stacks.
+		 * When the stack size exceeds `MAX_STACK_SIZE`, the oldest {@link Revertible} instance is disposed.
+		 * @param stack - The primary stack that the {@link Revertible} instance is being added to.
+		 * @param setstack - The setter function for the primary stack.
+		 */
+		function maintainStackSize(
+			stack: Revertible[],
+			setStack: React.Dispatch<React.SetStateAction<Revertible[]>>,
+		): void {
+			const MAX_STACK_SIZE = 5;
```

Comment:
This seems pretty small. Have we noticed any performance issues by tracking much more?

Reply:
Forgot to revert it back to 20. Will increase the limit. Thanks!

Reply:
Even 20 seems pretty small, IMO. Maybe not if the only things we can undo are task creations / deletions. But if we extended undo/redo coverage to include editing properties of tasks, then I think we need a much larger max.